xax/nn/metrics.py

```diff
@@ -0,0 +1,77 @@
+"""Norm and metric utilities."""
+
+from typing import Literal, cast, get_args
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array
+
+NormType = Literal["l1", "l2"]
+
+
+def cast_norm_type(norm: str) -> NormType:
+    if norm not in get_args(NormType):
+        raise ValueError(f"Invalid norm: {norm}")
+    return cast(NormType, norm)
+
+
+def get_norm(x: Array, norm: NormType) -> Array:
+    match norm:
+        case "l1":
+            return jnp.abs(x)
+        case "l2":
+            return jnp.square(x)
+        case _:
+            raise ValueError(f"Invalid norm: {norm}")
+
+
+def compute_distance_matrix(a: Array, b: Array) -> Array:
+    has_features = len(a.shape) > 1
+    a = jnp.expand_dims(a, axis=1)
+    b = jnp.expand_dims(b, axis=0)
```

Comment:
probably better to do this along last dimensions instead
```suggestion
    a = jnp.expand_dims(a, axis=-1)
    b = jnp.expand_dims(b, axis=-2)
```